### PR TITLE
references: update DM FAQ for 4GB binlog

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What is changed, added or deleted? 

update the FAQ for DM

### What is the related PR or file link(s)? 

update FAQ to handle the error occurred by larger binlog file (> 4GB)

### Which version does your change affect? 

dev, v3.0 and v2.1

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@WangXiangUSTC @lilin90 PTAL

File: dev/faq/data-migration.md [hunk 1]
```diff
@@ -20,9 +20,29 @@ category: FAQ
...
+原因是 DM 在写 relay log 时需要依据 binlog position 及文件大小对 event 进行验证，且需要保存同步的 binlog position 信息作为 checkpoint。但是 MySQL binlog position 官方定义使用 uint32 存储，所以超过 4G 部分的 binlog position 的 offset 值会溢出，进而出现上面的错误。
+
+对于 relay 处理单元，可通过以下步骤手动恢复：
+
+1. 在上游确认出错时对应的 binlog 文件的大小超出了 4GB。
+2. 停止 DM-worker。
+3. 将上游对应的 binlog 文件复制到 relay log 目录作为 relay log 文件。
+4. 更新 relay log 目录内对应的 _relay.meta_ 文件以从下一个 binlog 开始拉取。
+
+    如报错时有 `binlog-name = "mysql-bin.004451"` 与`binlog-pos = 2453`，则将其分别更新为 `binlog-name = "mysql-bin.004452"` 与`binlog-pos = 4`。
```
Comment (WangXiangUSTC) [inline dev/faq/data-migration.md]:
```suggestion
    例如：报错时有 `binlog-name = "mysql-bin.004451"` 与`binlog-pos = 2453`，则将其分别更新为 `binlog-name = "mysql-bin.004452"` 与`binlog-pos = 4`。
```

File: dev/faq/data-migration.md [hunk 2]
```diff
@@ -20,9 +20,29 @@ category: FAQ
...
+5. 重启 DM-worker。
+
+对于 binlog replication 处理单元，可通过以下步骤手动恢复：
+
+1. 在上游确认出错时对应的 binlog 文件的大小超出了 4GB。
+2. 通过 `stop-task` 停止同步任务。
+3. 将下游 `dm_meta` 数据库中 global checkpoint 与每个 table 的 checkpoint 中的 `binlog_name` 更新为出错的 binlog 文件、`binlog_pos` 更新为已同步过的一个合法的 position 值，比如 4。
```
Comment (WangXiangUSTC) [inline dev/faq/data-migration.md]: binlog name 是不是不用改？

File: dev/faq/data-migration.md [hunk 3]
```diff
@@ -20,9 +20,29 @@ category: FAQ
...
+
+    如出错任务名为 `dm_test`、对应的 `source-id` 为 `replica-1`，出错时对应的 binlog 文件为 `mysql-bin|000001.004451`，则执行 `UPDATE dm_test_syncer_checkpoint SET binlog_name='mysql-bin|000001.004451', binlog_pos = 4 WHERE id='replica-1';`。
```
Comment (WangXiangUSTC) [inline dev/faq/data-migration.md]:
```suggestion
    例如：出错任务名为 `dm_test`，对应的 `source-id` 为 `replica-1`，出错时对应的 binlog 文件为 `mysql-bin|000001.004451`，则执行 `UPDATE dm_test_syncer_checkpoint SET binlog_name='mysql-bin|000001.004451', binlog_pos = 4 WHERE id='replica-1';`。
```

File: dev/faq/data-migration.md [hunk 4]
```diff
@@ -20,9 +20,29 @@ category: FAQ
...
+4. 在同步任务配置中为 `sync` 部分设置 `safe-mode: true` 以保证可重入执行。
```
Comment (WangXiangUSTC) [inline dev/faq/data-migration.md]:
```suggestion
4. 在同步任务配置中为 `syncers` 部分设置 `safe-mode: true` 以保证可重入执行。
```

File: dev/faq/data-migration.md (hunk 2 above)
Comment (csuzhangxc) [reply dev/faq/data-migration.md]: 主要是会改 table 的 checkpoint（因为它也可能已经更新为超 4GB 溢出后的某个值）。而 table 的 checkpoint 可能 binlog name 还在更早的位置，所以想着统一更新一下

Comment (WangXiangUSTC) [reply dev/faq/data-migration.md]: ok

**Comment by WangXiangUSTC** [comment]
LGTM

File: dev/faq/data-migration.md [hunk 5]
```diff
@@ -20,9 +20,29 @@ category: FAQ
 
 当前版本 DM 发生该类型错误时，需要先使用 `stop-task` 停止任务后再使用 `start-task` 重启任务。后续 DM 会完善对此错误类型的自动重试机制。
 
-## 同步任务中断并包含 `get binlog error ERROR 1236 (HY000)`, `binlog checksum mismatch, data may be corrupted` 等 binlog 获取或解析失败错误
+## relay 处理单元报错 `event from * in * diff from passed-in event *` 或同步任务中断并包含 `get binlog error ERROR 1236 (HY000)`, `binlog checksum mismatch, data may be corrupted` 等 binlog 获取或解析失败错误
 
-在 DM 进行增量同步过程中，如果遇到了上游超过 4G 的 binlog 文件，并且 DM 在处理这个 binlog 过程中出现了同步中断（包括正常的 pause/stop 任务，异常的原因导致的中断时）, 就可能出现这个错误。原因是 DM 需要保存同步的 binlog position 信息，但是 MySQL binlog position 官方定义使用 uint32 存储，所以超过 4G 部分的 binlog position 的 offset 值会溢出，就会存储的是一个错误的 binlog position，在重启 task 或者 dm-worker 后，需要使用该 binlog position 重新解析 binlog/relay log，进而出现上面的错误。遇到这种情况需要手动进行恢复，恢复方法：
+在 DM 进行 relay log 拉取与增量同步过程中，如果遇到了上游超过 4GB 的 binlog 文件，就可能出现这个错误。
```
Comment (GregoryIan) [inline dev/faq/data-migration.md]: 就可能出现这两个错误？

File: dev/faq/data-migration.md [hunk 6]
```diff
@@ -20,9 +20,29 @@ category: FAQ
...
+对于 binlog replication 处理单元，可通过以下步骤手动恢复：
+
+1. 在上游确认出错时对应的 binlog 文件的大小超出了 4GB。
+2. 通过 `stop-task` 停止同步任务。
+3. 将下游 `dm_meta` 数据库中 global checkpoint 与每个 table 的 checkpoint 中的 `binlog_name` 更新为出错的 binlog 文件、`binlog_pos` 更新为已同步过的一个合法的 position 值，比如 4。
+
+    例如：出错任务名为 `dm_test`，对应的 `source-id` 为 `replica-1`，出错时对应的 binlog 文件为 `mysql-bin|000001.004451`，则执行 `UPDATE dm_test_syncer_checkpoint SET binlog_name='mysql-bin|000001.004451', binlog_pos = 4 WHERE id='replica-1';`。
+4. 在同步任务配置中为 `syncers` 部分设置 `safe-mode: true` 以保证可重入执行。
+5. 通过 `start-task` 启动同步任务。
+6. 通过 `query-status` 观察同步任务状态，当原造成出错的 relay log 文件同步完成后，即可关闭 `safe-mode` 并重启同步任务。
```
Comment (GregoryIan) [inline dev/faq/data-migration.md]: 还原 `safe-mode` 为原始值

File: dev/faq/data-migration.md (hunk 5 above)
Comment (csuzhangxc) [reply dev/faq/data-migration.md]: done in https://github.com/pingcap/docs-cn/pull/1924/commits/423ecdc056471c86143706a89bcf8a8d4cb36452.

File: dev/faq/data-migration.md (hunk 6 above)
Comment (csuzhangxc) [reply dev/faq/data-migration.md]: done in https://github.com/pingcap/docs-cn/pull/1924/commits/423ecdc056471c86143706a89bcf8a8d4cb36452.

File: dev/faq/data-migration.md [hunk 7]
```diff
@@ -20,9 +20,29 @@ category: FAQ
 
 当前版本 DM 发生该类型错误时，需要先使用 `stop-task` 停止任务后再使用 `start-task` 重启任务。后续 DM 会完善对此错误类型的自动重试机制。
 
-## 同步任务中断并包含 `get binlog error ERROR 1236 (HY000)`, `binlog checksum mismatch, data may be corrupted` 等 binlog 获取或解析失败错误
+## relay 处理单元报错 `event from * in * diff from passed-in event *` 或同步任务中断并包含 `get binlog error ERROR 1236 (HY000)`, `binlog checksum mismatch, data may be corrupted` 等 binlog 获取或解析失败错误
```
Comment (lilin90) [inline dev/faq/data-migration.md]:
```suggestion
## relay 处理单元报错 `event from * in * diff from passed-in event *` 或同步任务中断并包含 `get binlog error ERROR 1236 (HY000)`、`binlog checksum mismatch, data may be corrupted` 等 binlog 获取或解析失败错误
```

File: dev/faq/data-migration.md [hunk 8]
```diff
@@ -20,9 +20,29 @@ category: FAQ
...
-- 首先判断出错发生在 relay log 写入还是 binlog replication/syncer unit 同步（通过日志出错信息中的 component 信息即可判断），如果错误发生在 relay log 模块，binlog replication/syncer unit 保存的断点都是正确的情况，可以先停止任务，停止 DM-worker，手动调节 relay meta 的 binlog-position 到 4，重启 DM-worker 重新拉取 relay log ，relay log 写入正常后启动任务会自动从断点继续同步。
-- 如果 relay log 写入正常，已经写到了下一个文件，错误发生在 binlog replication/syncer unit 读取当前超过 4G relay log 文件的一个不合法的 binlog position，这时候可以停止任务，手动调节该任务的 binlog position 信息到该 relay log 的一个合法的 binlog position，比如 4，注意需要同时调整 global checkpoint 和每个 table checkpoint 的 binlog position。设置任务的 safe-mode 为 true ，保证可重入执行，在这之后就可以重新启动同步任务，观察同步状态，只要同步过这个大于 4G 的文件即可。
+原因是 DM 在写 relay log 时需要依据 binlog position 及文件大小对 event 进行验证，且需要保存同步的 binlog position 信息作为 checkpoint。但是 MySQL binlog position 官方定义使用 uint32 存储，所以超过 4G 部分的 binlog position 的 offset 值会溢出，进而出现上面的错误。
+
+对于 relay 处理单元，可通过以下步骤手动恢复：
+
+1. 在上游确认出错时对应的 binlog 文件的大小超出了 4GB。
+2. 停止 DM-worker。
+3. 将上游对应的 binlog 文件复制到 relay log 目录作为 relay log 文件。
+4. 更新 relay log 目录内对应的 _relay.meta_ 文件以从下一个 binlog 开始拉取。
```
Comment (lilin90) [inline dev/faq/data-migration.md]: We use the inline code format for file names to make it easier to detect.

```suggestion
4. 更新 relay log 目录内对应的 `relay.meta` 文件以从下一个 binlog 开始拉取。
```

File: dev/faq/data-migration.md [hunk 9]
```diff
@@ -20,9 +20,29 @@ category: FAQ
...
+
+    例如：报错时有 `binlog-name = "mysql-bin.004451"` 与`binlog-pos = 2453`，则将其分别更新为 `binlog-name = "mysql-bin.004452"` 与`binlog-pos = 4`。
+5. 重启 DM-worker。
+
+对于 binlog replication 处理单元，可通过以下步骤手动恢复：
+
+1. 在上游确认出错时对应的 binlog 文件的大小超出了 4GB。
+2. 通过 `stop-task` 停止同步任务。
+3. 将下游 `dm_meta` 数据库中 global checkpoint 与每个 table 的 checkpoint 中的 `binlog_name` 更新为出错的 binlog 文件、`binlog_pos` 更新为已同步过的一个合法的 position 值，比如 4。
```
Comment (lilin90) [inline dev/faq/data-migration.md]:
```suggestion
3. 将下游 `dm_meta` 数据库中 global checkpoint 与每个 table 的 checkpoint 中的 `binlog_name` 更新为出错的 binlog 文件，将 `binlog_pos` 更新为已同步过的一个合法的 position 值，比如 4。
```

File: dev/faq/data-migration.md [hunk 10]
```diff
@@ -20,9 +20,29 @@ category: FAQ
...
+
+    例如：出错任务名为 `dm_test`，对应的 `source-id` 为 `replica-1`，出错时对应的 binlog 文件为 `mysql-bin|000001.004451`，则执行 `UPDATE dm_test_syncer_checkpoint SET binlog_name='mysql-bin|000001.004451', binlog_pos = 4 WHERE id='replica-1';`。
+4. 在同步任务配置中为 `syncers` 部分设置 `safe-mode: true` 以保证可重入执行。
```
Comment (lilin90) [inline dev/faq/data-migration.md]: How to understand `重入执行`?

Comment (csuzhangxc) [reply dev/faq/data-migration.md]: `reentrant`

Comment (WangXiangUSTC) [reply dev/faq/data-migration.md]: 我们后面在专有名词里解释一下吧

Comment (lilin90) [reply dev/faq/data-migration.md]: Got it.

**Pull request merged**